Linenumber directives from markdown files (#5)
driusan/lmt: LineNumbers.md

# Linenumber directives

Most languages have linenumber directives. They are used for generated code so

> debuggers and errormessages are able to tell which line from the source file is

Reviewer: ```suggestion
debuggers and error messages are able to tell which line from the source file is
```

> responsible instead of the line from the intermediate representation. Golang

Reviewer: ```suggestion
responsible instead of the line from the intermediate representation. Go
```

uses '//line [filename]:[linenumber]', where [XXX] denotes an optional

> parameter. Many (most?) languages from the C-family instead used the format

Reviewer: ```suggestion
parameter. Many languages from the C-family instead use the format
```

> '#line lineno ["filename"]'.

Reviewer: ```suggestion
`#line lineno ["filename"]`.
```

> This is quite usefule, since `go build`, `go run` etc will provide references

Reviewer: ```suggestion
This is quite useful, since `go build`, `go run` etc will provide references
```

to the markdown file with the bug (well, at least in theory). IF we want to
debug the intermediate file, a reference above the specific code is always
readily avialible a quick right click (Acme) or `gF` away (vim).

If we want to implement this for lmt, we need to change quite a few things, the
most obvious is the internal representation of a line, which needs quite a lot

> of more metadata, as the filename, the linenumber and the language of the

Reviewer: ```suggestion
more metadata such as the filename, the line number and the language of the
```

codeblock.

> We remake CodeBlock to a slice of CodeLines. For every line we don't only

Reviewer: ```suggestion
We need to change CodeBlock to a slice of CodeLines. For every line we don't only
```

record the text but the name of the source file, the linenumber and the
language (for different line directives).

```go "global block variables"
type File string
type CodeBlock []CodeLine
type BlockName string
type language string
type CodeLine struct {
	text   string
	file   File
	lang   language
	number int
}

var blocks map[BlockName]CodeBlock
var files map[File]CodeBlock
```

Handling a block line is now not as simple as just joining two strings. We
start appending CodeLines into CodeBlocks.

```go "Handle block line"
block = append(block, line)
```

This cascades. The CodeBlocks in the maps files and blocks can't be joined, but

> has to be appended. We take a few moments to clean up the emptying of metadata

Reviewer: ```suggestion
have to be appended. We take a few moments to clean up the emptying of metadata
```

per line, i.e. if a block is ending we can unset inBlock.

```go "Handle block ending"
inBlock = false
// Update the files map if it's a file.
if fname != "" {
	if appending {
		files[fname] = append(files[fname], block...)
	} else {
		files[fname] = block
	}
}

// Update the named block map if it's a named block.
if bname != "" {
	if appending {
		blocks[bname] = append(blocks[bname], block...)
	} else {
		blocks[bname] = block
	}
}
```

To get this to work we need to start using CodeLine as line in our
implementation of the file processing. We also should to save the filename
directly when opening the file. To simplify we break out our variables from the
implementation of file processing.

```go "process file implementation variables"
scanner := bufio.NewReader(r)
var err error

var line CodeLine
line.file = File(inputfilename)

var inBlock, appending bool
var bname BlockName
var fname File
var block CodeBlock
```

When processing our files we need to increment the file line counter for every
line read.

```go "process file implementation"
<<<process file implementation variables>>>
for {

> 	line.text, err = scanner.ReadString('\n')

Reviewer: not sure if it's just how GitHub is showing it or not, but the indentation here looks wrong. It should be tabs. You can highlight the code block in ie. acme and pipe that section through `gofmt` to make sure the formatting of theGo code is formatted correctly. (If there are <<< .... >>> macros as a hack you can temporarily comment them, use gofmt, then uncomment them)

	line.number++
	switch err {
	case io.EOF:
		return nil
	case nil:
		// Nothing special
	default:
		return err
	}
	<<<Handle file line>>>
}
```

> Sadly, there is no indicator from ioreader which file we are reading so we send

Reviewer: ```suggestion
Sadly, there is no indicator from the Reader about which file we are reading so we send
```

> it in with ProcessFile.

Reviewer: ```suggestion
it along with ProcessFile.
```

if matches == nil {
	if v.text != "\n" {
		v.text = prefix + v.text
	}
	ret = append(ret, v)
	continue
}
<<<Lookup replacement and add to ret>>>
```

Finally something we can replace with a simpler implementation. Instead of
reading line by line from c (a CodeBlock, earlier a long string with newlines),
we're now able to use range. The string `line` is still analyzed, "de-macro-ed"
and prefixed.

```go "Replace codeblock implementation"
var line string
for _, v := range c {
	line = v.text
	<<<Handle replace line>>>
}
return
```

We need to use all the information we gathered, lets add a "Finalize"-function
which handles CodeLines and outputs the final strings with the line-directives
we been working for.

```go "other functions" +=
<<<Finalize Declaration>>>
```

Let us implement a method which acts upon a codeblock and returns a string
containing the ready to be used textual representation of the code (or other
files). We check if the filename is the same as the previous line, and if the
linenumber has been increased by exactly one. If any of those have changed it
is because the source for this line are not the same as the previous and we

> interject a "line directive". We know of two variants of these go and the

Reviewer: ```suggestion
interject a "line directive". We know of two variants of these: Go and the
```

> common from the C family of languages. lastly we save state, so we have

Reviewer: ```suggestion
 C variant. Lastly we save state, so we have
```

something to compare to on the next line.

```go "Finalize Declaration"

// Finalize extract the textual lines from CodeBlocks and (if needed) prepend a
// notice about "unexpected" filename or line changes, which is extracted from
// the contained CodeLines. The result is a string with newlines ready to be
// pasted into a file.
func (c CodeBlock) Finalize() (ret string) {
	var file File
	var formatstring string
	var linenumber int
	for _, l := range c {
		if linenumber+1 != l.number || file != l.file {
			switch l.lang {
			case "go", "golang":
				formatstring = "//line %[2]v:%[1]v\n"

> 			default:

Reviewer: ```suggestion
			case "C", "c":
```

I don't think it makes sense to include a global default, it's not a safe assumption that if it's not Go it's C. It's better to just not include line number if it's not a known language.
